sql: clean up uses of Statement

RaduBerinde opened the pull request:
This change cleans up the use of `sql.Statement` and reduces some
allocations. Specifically:

 - we create a  `Statement` lower in the stack (in `execStmtInOpenState`),
   and pass only what we need in the higher layers;

 - we change various functions to take a `tree.Statement` rather than
   an entire `Statement` when possible;

 - we move down the `withStatement` context allocation, so that it is
   avoided in the implicit transaction state transition;

 - we store a copy rather than a pointer to the Statement in the
   planner;

 - we avoid directly using `stmt` fields from `func()` declarations
   that escape;

 - we populate `Statement.AnonymizedStr` upfront. The anonymized
   string is always needed (to update statement stats).

```
name                               old time/op    new time/op    delta
EndToEnd/kv-read/EndToEnd             153µs ± 1%     154µs ± 2%    ~     (p=0.486 n=4+4)
EndToEnd/kv-read-no-prep/EndToEnd     216µs ± 1%     217µs ± 1%    ~     (p=0.886 n=4+4)
EndToEnd/kv-read-const/EndToEnd       111µs ± 1%     113µs ± 1%  +1.01%  (p=0.029 n=4+4)

name                               old alloc/op   new alloc/op   delta
EndToEnd/kv-read/EndToEnd            25.8kB ± 1%    25.5kB ± 1%    ~     (p=0.114 n=4+4)
EndToEnd/kv-read-no-prep/EndToEnd    32.2kB ± 1%    31.9kB ± 1%    ~     (p=0.686 n=4+4)
EndToEnd/kv-read-const/EndToEnd      21.0kB ± 1%    20.7kB ± 2%    ~     (p=0.200 n=4+4)

name                               old allocs/op  new allocs/op  delta
EndToEnd/kv-read/EndToEnd               252 ± 1%       250 ± 0%  -0.99%  (p=0.029 n=4+4)
EndToEnd/kv-read-no-prep/EndToEnd       332 ± 0%       330 ± 1%    ~     (p=0.229 n=4+4)
EndToEnd/kv-read-const/EndToEnd         214 ± 0%       212 ± 0%  -0.93%  (p=0.029 n=4+4)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56217)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but it'd probably be beneficial for someone else to also have a look.

> Reviewed 18 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56217#-:-MLA0ykNDLiKexPmU_7Y:bxl6hvz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @nvanbenschoten, and @RaduBerinde)

___
*[pkg/sql/conn_executor_exec.go, line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56217#-MLA-R8L1CA9uvPatIdY:-MLA-R8MCVbqZ7XGOXsk:b-b9j9ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6c18541f182aec0ed3ed6bfa63398f97116745/pkg/sql/conn_executor_exec.go#L203)):*
> ```Go
> 		// StatementType() different from "Rows".
> 		if portal.exhausted {
> 			return nil, nil, nil
> ```

nit: not sure if it's important, but maybe we should just have `return` without `nil`s so that default `ev, payload, err` values are used?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

:100:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56217#-:-MLA9Wma6nGEPvIho9T7:birsy2e)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @nvanbenschoten and @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56217#-MLA-R8L1CA9uvPatIdY:-MLA9NPg8PdvgU57d_II:b-vgnrwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea6c18541f182aec0ed3ed6bfa63398f97116745/pkg/sql/conn_executor_exec.go#L203)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: not sure if it's important, but maybe we should just have `return` without `nil`s so that default `ev, payload, err` values are used?
</blockquote></details>

no... no naked returns pls :)


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde committed:
sql: clean up uses of Statement

This change cleans up the use of `sql.Statement` and reduces some
allocations. Specifically:

 - we create a  `Statement` lower in the stack (in execStmtInOpenState),
   and pass only what we need in the higher layers;

 - we change various functions to take a `tree.Statement` rather than
   an entire Statement when possible;

 - we move down the withStatement context allocation, so that it is
   avoided in the implicit transaction state transition;

 - we store a copy rather than a pointer to the Statement in the
   planner;

 - we avoid directly using `stmt` fields from `func()` declarations
   that escape;

 - we populate `Statement.AnonymizedStr` upfront. The anonymized
   string is always needed (to update statement stats).

```
name                               old time/op    new time/op    delta
EndToEnd/kv-read/EndToEnd             153µs ± 1%     154µs ± 2%    ~     (p=0.486 n=4+4)
EndToEnd/kv-read-no-prep/EndToEnd     216µs ± 1%     217µs ± 1%    ~     (p=0.886 n=4+4)
EndToEnd/kv-read-const/EndToEnd       111µs ± 1%     113µs ± 1%  +1.01%  (p=0.029 n=4+4)

name                               old alloc/op   new alloc/op   delta
EndToEnd/kv-read/EndToEnd            25.8kB ± 1%    25.5kB ± 1%    ~     (p=0.114 n=4+4)
EndToEnd/kv-read-no-prep/EndToEnd    32.2kB ± 1%    31.9kB ± 1%    ~     (p=0.686 n=4+4)
EndToEnd/kv-read-const/EndToEnd      21.0kB ± 1%    20.7kB ± 2%    ~     (p=0.200 n=4+4)

name                               old allocs/op  new allocs/op  delta
EndToEnd/kv-read/EndToEnd               252 ± 1%       250 ± 0%  -0.99%  (p=0.029 n=4+4)
EndToEnd/kv-read-no-prep/EndToEnd       332 ± 0%       330 ± 1%    ~     (p=0.229 n=4+4)
EndToEnd/kv-read-const/EndToEnd         214 ± 0%       212 ± 0%  -0.93%  (p=0.029 n=4+4)
```

Release note: None

RaduBerinde commented:
bors r+
